### PR TITLE
add prod

### PR DESCRIPTION
**Pull request by oxinabox**
I wanted to see if it was as simple as i thought it would be.
The answer is yes.

### CODE REVIEW

File: test/rulesets/Base/mapreduce.jl [hunk 1]
```diff
@@ -20,4 +20,11 @@
             end
         end
     end  # sum abs2
+
+    @testset "prod" begin
+        test_rrule(prod, randn(5))
+        test_rrule(prod, randn(5, 6))
+        test_rrule(prod, randn(5, 6); fkwargs=(;dims=2))
+        test_rrule(prod, randn(5, 6); fkwargs=(;dims=1))
```
Comment (simeonschaub) [inline test/rulesets/Base/mapreduce.jl:28]: Needs complex tests as well

File: src/rulesets/Base/mapreduce.jl [hunk 2]
```diff
@@ -48,3 +48,17 @@ function rrule(
     end
     return y, sum_abs2_pullback
 end
+
+####
+#### prod
+####
+
+function rrule(::typeof(prod), x::AbstractArray{T}; dims=:) where {T<:CommutativeMulNumber}
```
Comment (willtebbutt) [inline src/rulesets/Base/mapreduce.jl:56]: Would you consider restricting the type of `x`?

Comment (oxinabox) [reply src/rulesets/Base/mapreduce.jl:56]: I just matched `sum`

Comment (willtebbutt) [reply src/rulesets/Base/mapreduce.jl:56]: Given that we've not tied this down, I propose `StridedArray`.

**Comment by sethaxen** [comment]
Can you comment on how this is relates to https://github.com/JuliaDiff/ChainRules.jl/pull/335? Does this cover the same cases?

**Comment by oxinabox** [comment]
I was not aware of #335
This is definitely simpler that #335.
So I guess I am missing something

**Comment by oxinabox** [comment]
Yep the following fails:
```
julia>         test_rrule(prod, [2.0, 0.0, 1.0, 3.5])
test_rrule: prod at ([2.0, 0.0, 1.0, 3.5],): Test Failed at /Users/oxinabox/.julia/packages/ChainRulesTestUtils/AIKXr/src/check_result.jl:19
  Expression: isapprox(actual, expected; kwargs...)
   Evaluated: isapprox([0.0, NaN, 0.0, 0.0], [0.0, 0.8173013508577128, 0.0, 0.0]; rtol = 1.0e-9, atol = 1.0e-9)
Stacktrace:
 [1] check_equal(::Array{Float64,1}, ::Array{Float64,1}; kwargs::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol,Symbol},NamedTuple{(:rtol, :atol),Tuple{Float64,Float64}}}) at /Users/oxinabox/.julia/packages/ChainRulesTestUtils/AIKXr/src/check_result.jl:19
 [2] macro expansion at /Users/oxinabox/.julia/packages/ChainRulesTestUtils/AIKXr/src/testers.jl:195 [inlined]
 [3] macro expansion at /usr/local/src/julia/julia-1.5/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115 [inlined]
 [4] test_rrule(::typeof(prod), ::Array{Float64,1}; output_tangent::ChainRulesTestUtils.Auto, fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5,1,FiniteDifferences.UnadaptedFiniteDifferenceMethod{7,5}}, check_inferred::Bool, fkwargs::NamedTuple{(),Tuple{}}, rtol::Float64, atol::Float64, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /Users/oxinabox/.julia/packages/ChainRulesTestUtils/AIKXr/src/testers.jl:160
 ```
 
 Closing this infavor of #335

**Pull request closed without merging**